Cython/TestUtils.py

```diff
@@ -228,21 +229,42 @@ def unpack_source_tree(tree_file, workdir, cython_root):
     return workdir, header
 
 
+def fresh_cythonize(*args, **kwargs):
+    """Clear function caches and run cythonize"""
+    clear_function_and_Dependencies_caches()
+    Dependencies.cythonize(*args, **kwargs)
+
+
+def clear_function_and_Dependencies_caches():
+    Utils.clear_function_caches()
+    Dependencies._dep_tree = None  # discard method caches
+
+
 def write_file(file_path, content, dedent=False):
     """
-    Write some content (text or bytes) to the file at "file_path".
+    Write some content (text or bytes) to the file at `file_path` in utf-8.
     """
-    mode = 'wb' if isinstance(content, bytes) else 'w'
+
+    kwargs = {}
+
+    mode = "w"
+    if isinstance(content, bytes):
+        mode += "b"
+    else:
+        # unlike OS X or Linux,
+        # Windows defaults to 8-bit character set like windows-1252
+        kwargs["encoding"] = "utf-8"
+
     if dedent:
         content = textwrap.dedent(content)
 
-    with open(file_path, mode=mode) as f:
+    with open(file_path, mode=mode, **kwargs) as f:
```

Comment:
This can be simplified by setting `encoding = None` and always passing it here. In fact, `encoding='utf-8'` should be a parameter of the two helper functions, which is only ignored for `bytes` input.

Reply:
> In fact, `encoding='utf-8'` should be a parameter of the two helper functions, which is only ignored for `bytes` input.

YAGNI?

Reply:
Are you sure you want to use something other than utf-8?

Reply:
[resolved](https://github.com/cython/cython/pull/4241/commits/4e33d3168370d28653e75139c7a7af80615172d5)

Reply:
>> This can be simplified by setting `encoding = None` and always passing it here. In fact, `encoding='utf-8'` should be a parameter of the two helper functions, which is only ignored for `bytes` input.
>
> YAGNI ?

This isn't about adding a new feature. You have to put the default somewhere. If I have to decide between putting it into a code line inside of the function body here and making it a default argument of the function, then I don't see why we shouldn't make it a default argument. It totally seems like something that we may want to pass in.

> Are you sure you want to use something other than utf-8?

Doesn't seem unreasonable to me, give that this is a compiler project that processes source code files. ;-)

Reply:
And yes, one more thing, if you want to continue the conversation, it would be more convenient if you open it again. If you are just commenting and do not want me to change anything, then you can leave the conversation resolved.

Reply:
I unresolved it so it easier to see.

Reply:
Just to progress in this PR, and not argue about YAGNI, I implemented it [here](https://github.com/cython/cython/pull/4241/commits/c45b37e0a8469de641cbcc479410a5bacce0f836)